File: C++/0003-longest-substring-without-repeating-characters.cpp
class Solution {
 public:
  int lengthOfLongestSubstring(string s) {
    vector<int> chars(128);
    int left = 0, right = 0, res = 0;
    while (right < s.length()) {
      chars[s[right]]++;
      while (chars[s[right]] > 1) {
        chars[s[left]]--;
        left++;
      }
      res = max(res, right - left + 1);
      right++;
    }
    return res;
  }
};

File: Python/0003-longest-substring-without-repeating-characters.py
# time complexity: O(n)
# space complexity: O(min(m,n))
from collections import Counter


class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        chars = Counter()
        left = right = 0
        res = 0
        while right < len(s):
            r = s[right]
            chars[r] += 1
            while chars[r] > 1:
                l = s[left]
                chars[l] -= 1
                left += 1
            res = max(res, right - left + 1)
            right += 1
        return res


Input = "abcabcbb"
print(Solution().lengthOfLongestSubstring(Input))
